### PR TITLE
feat: Add ClickHouse driver to sql inputs/outputs plugins

### PR DESCRIPTION
**Pull request by nahsi**
### Required for all PRs:

- [x] Updated associated README.md.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary) (e.g. feat: or fix:)

Added ClickHouse to sql input drivers (using [clickhouse-go](https://github.com/ClickHouse/clickhouse-go))

----

Since there are no changes to plugin code, I'm not sure there is a need to add tests.

### CODE REVIEW

**Comment by nahsi** [comment]
!signed-cla

**Review by srebhan — APPROVED** [review]
Looks good to me! Thanks for adding this driver @nahsi!

**Comment by srebhan** [comment]
@nahsi one additional request, can you please also prepare a pull-request for `output/sql` so we can keep the two symmetrical...

**Comment by srebhan** [comment]
@nahsi and another request: Please add a test-case, i.e. a docker instance where you can import the example tables. Take a look at the postgres and mysql tests in the repo.

**Comment by nahsi** [comment]
@srebhan to use clickhouse-go client I had to use `begin->prepare->exec->commit` instead of just `exec` in outputs/sql.

With just `exec` clickhouse-go will emit [this](https://github.com/ClickHouse/clickhouse-go/blob/master/clickhouse.go#L36) error.

I'm not sure if it is OK since I don't know golang and I know clickhouse only from OPS point of view :) But it works.

Maybe it would be better to `begin` then loop over metrics preparing sql etc and then `commit`? I don't know. If it is so I could try to update.

**Comment by srebhan** [comment]
@nahsi any update on this PR?

**Comment by nahsi** [comment]
Sorry, I will do it this week.

**Comment by nahsi** [comment]
@srebhan I've commited your suggestions, ready for another review.

File: plugins/outputs/sql/sql.go [hunk 1]
```diff
@@ -248,15 +251,36 @@ func (p *SQL) Write(metrics []telegraf.Metric) error {
...
-		if err != nil {
-			// check if insert error was caused by column mismatch
-			p.Log.Errorf("Error during insert: %v, %v", err, sql)
-			return err
+		switch p.Driver {
+		case "clickhouse":
+			// ClickHouse needs to batch inserts with prepared statements
+			tx, err := p.db.Begin()
+			if err != nil {
+				return fmt.Errorf(": %v", err)
```
Comment (srebhan) [inline plugins/outputs/sql/sql.go:260]: There is something missing I guess... Maybe
```suggestion
				return fmt.Errorf("begin failed: %v", err)
```

File: plugins/outputs/sql/sql.go [hunk 2]
```diff
@@ -248,15 +251,36 @@ func (p *SQL) Write(metrics []telegraf.Metric) error {
...
+			}
+			stmt, err := tx.Prepare(sql)
+			if err != nil {
+				return fmt.Errorf("Error during prepare: %v", err)
```
Comment (srebhan) [inline plugins/outputs/sql/sql.go:264]: Errors in Golang [should start lowercase without trailing punctuation](https://github.com/golang/go/wiki/CodeReviewComments#error-strings). How about
```suggestion
				return fmt.Errorf("prepare failed: %v", err)
```
Same for the other errors below.

File: plugins/outputs/sql/sql.go [hunk 3]
```diff
@@ -248,15 +251,36 @@ func (p *SQL) Write(metrics []telegraf.Metric) error {
...
+			}
+			defer stmt.Close() //nolint
```
Comment (srebhan) [inline plugins/outputs/sql/sql.go:266]: Please limit the nolint scope. I guess
```suggestion
			//nolint:revive // We cannot do anything about a failing close.
			defer stmt.Close()
```
will do.

File: plugins/outputs/sql/sql.go [hunk 4]
```diff
@@ -248,15 +251,36 @@ func (p *SQL) Write(metrics []telegraf.Metric) error {
...
+			}
+		default:
+			_, err = p.db.Exec(sql, values...)
+			if err != nil {
+				return fmt.Errorf("Error during execution: %v", err)
+			}
 		}
 	}
-	return nil
+	return err
```
Comment (srebhan) [inline plugins/outputs/sql/sql.go:283]: I think we can only end up here if there is no error so keeping `return nil` is fine.

**Review by srebhan — COMMENTED** [review]
@nahsi thanks for the update. There are some minor formatting issues, but nothing too big.

File: plugins/outputs/sql/sql.go (hunk 2 above)
Comment (nahsi) [reply plugins/outputs/sql/sql.go:264]: I see, I didn't know about that. Thank you for the link!

**Comment by nahsi** [comment]
@srebhan I've commited the changes you suggested. Please have a look again when you have time

**Review by srebhan — APPROVED** [review]
@nahsi perfect thank you for your contribution!

Looks good to me.

**Review by powersj — APPROVED** [review]

**Comment by crabvk** [comment]
Maybe I'm doing something wrong, but I'm getting this error
> 2022-02-14T05:32:39Z E! [agent] Failed to connect to [outputs.sql], retrying in 15s, error was 'code: 516, message: default: Authentication failed: password is incorrect or there is no user with such name'

Using following telegraf config
```toml
[[outputs.sql]]
  driver = "clickhouse"
  data_source_name = "clickhouse://user:password@example.com:9000/database"
  [outputs.sql.convert]
    integer = "Int64"
    text = "String"
    timestamp = "DateTime"
    defaultvalue = "String"
    unsigned = "UInt64"
    bool = "Uint8"
```
However, from the same server I'm able to connect to clickhouse using cli
```shell
clickhouse-client --user user --password password --host example.com --database database
```
I've checked credentials several times, they are the same in both cases.

**Comment by crabvk** [comment]
Oops, I was using wrong connection string from clickhouse-go v2, for v1 it has different format
```toml
data_source_name = "tcp://example.com:9000?username=user&password=password&database=database"
```
Now it works.
Thank you for the plugin 🙂

**Comment by crabvk** [comment]
Got another error now
```log
2022-02-14T06:01:11Z E! [agent] Error writing to outputs.sql: code: 62, message: Syntax error: failed at position 72 ('UInt64'): UInt64,"inactive" Int64 UInt64,"vmalloc_chunk" Int64 UInt64,"sreclaimable" Int64 UInt64,"swap_cached" Int64 UInt64,"used" Int64 UInt64,"huge_pages_free" Int64 U. Expected one of: DEFAULT, MATERIALIZED, ALIAS, NOT, NULL, COMMENT, CODEC, TTL, token, Comma, ClosingRoundBracket
```

<details>
  <summary>Clickhouse log message</summary>
2022.02.14 06:07:46.082356 [ 2196388 ] {1ff9aa5d-c0c5-4322-8b15-5e94806715ef} <Error> TCPHandler: Code: 62. DB::Exception: Syntax error: failed at position 72 ('UInt64'): UInt64,"inactive" Int64 UInt64,"vmalloc_chunk" Int64 UInt64,"sreclaimable" Int64 UInt64,"swap_cached" Int64 UInt64,"used" Int64 UInt64,"huge_pages_free" Int64 U. Expected one of: DEFAULT, MATERIALIZED, ALIAS, NOT, NULL, COMMENT, CODEC, TTL, token, Comma, ClosingRoundBracket. (SYNTAX_ERROR), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0xa82d07a in /usr/bin/clickhouse
1. DB::parseQueryAndMovePosition(DB::IParser&, char const*&, char const*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool, unsigned long, unsigned long) @ 0x14d642bf in /usr/bin/clickhouse
2. ? @ 0x13d121f0 in /usr/bin/clickhouse
3. DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum) @ 0x13d11e55 in /usr/bin/clickhouse
4. DB::TCPHandler::runImpl() @ 0x147f050f in /usr/bin/clickhouse
5. DB::TCPHandler::run() @ 0x14804259 in /usr/bin/clickhouse
6. Poco::Net::TCPServerConnection::start() @ 0x1745e52f in /usr/bin/clickhouse
7. Poco::Net::TCPServerDispatcher::run() @ 0x17460981 in /usr/bin/clickhouse
8. Poco::PooledThread::run() @ 0x17611609 in /usr/bin/clickhouse
9. Poco::ThreadImpl::runnableEntry(void*) @ 0x1760ed00 in /usr/bin/clickhouse
10. ? @ 0x7f5257a61259 in ?
11. clone @ 0x7f525798a5e3 in ?
</details>

ClickHouse version 22.1.3.7

**Comment by crabvk** [comment]
I think the cause of the problem is [here](https://github.com/influxdata/telegraf/blob/3b980ee0703ff598f17ec5ecd484408b3463c73b/plugins/outputs/sql/sql.go#L103), when value type is `uint64` clickhouse type becomes `Int64 UInt64` 😬

**Comment by srebhan** [comment]
@crabvk can you please open an _Issue_ for your _issue_ ;-) instead of putting the information next to this PR!?

**Comment by Okaho** [comment]
its worked for me:

[[outputs.sql]]
   driver = "clickhouse"
   data_source_name = "tcp://example.com:9000?username=user&password=password&database=database"
   timestamp_column = "timestamp"
   table_template = "CREATE TABLE IF NOT EXISTS {TABLE}({COLUMNS}) ENGINE = MergeTree() PRIMARY KEY (timestamp,host) ORDER by (timestamp,host) Partition by toYYYYMMDD(timestamp)"
    table_exists_template = "SELECT 1 FROM {TABLE} LIMIT 1"
  [outputs.sql.convert]
     integer              = "INT"
     real                 = "DOUBLE"
     text                 = "TEXT"
     timestamp            = "TIMESTAMP"
     defaultvalue         = "TEXT"
     unsigned             = "UNSIGNED"
     bool                 = "BOOL"

**Comment by zentkhv** [comment]
> 

Can you please show you full telegraf config from you project?

**Comment by Okaho** [comment]
Hi. What exactly interests you in telegrapr configuration?  Something is
wrong with CH?

вт, 21 февр. 2023 г., 10:31 zentkhv ***@***.***>:

**Comment by zentkhv** [comment]
HI!
Im trying to send metrics to clickhouse by telegraf with you part config example, but geting an error:
[agent] Failed to connect to [outputs.sql], retrying in 15s, error was 'driver: bad connection'
Add my full config.
I thought your full config will help me to solve my problem.

**Comment by Okaho** [comment]
```
telegraf --version
Telegraf 1.24.0 (git: HEAD@3c4a6516)
```

/etc/telegraf/telegraf.conf
```
[agent]
  interval = "60s"
  metric_batch_size = 100
  metric_buffer_limit = 10000
  collection_jitter = "0s"
  flush_interval = "1s"
  flush_jitter = "0s"
  precision = "0s"
  logfile = "/var/log/telegraf/telegraf.log"

  hostname = "host"
  omit_hostname = false

[[outputs.sql]]
   driver = "clickhouse"

data_source_name = "tcp://aa.bb.cc.dd:9000?username=username&password=password&database=telegraf3"

   timestamp_column = "timestamp"
   table_template = "CREATE TABLE IF NOT EXISTS {TABLE}({COLUMNS}) ENGINE = MergeTree() PRIMARY KEY (timestamp,host) ORDER by (timestamp,host) Partition by toYYYYMMDD(timestamp)"
    table_exists_template = "SELECT 1 FROM {TABLE} LIMIT 1"
  [outputs.sql.convert]
     integer              = "INT"
     real                 = "DOUBLE"
     text                 = "TEXT"
     timestamp            = "TIMESTAMP"
     defaultvalue         = "TEXT"
     unsigned             = "UNSIGNED"
     bool                 = "BOOL"


[[inputs.cpu]]
  percpu = true
  totalcpu = true
  collect_cpu_time = false
  report_active = false

[[inputs.disk]]
   mount_points = ["/"]
  ignore_fs = ["tmpfs", "devtmpfs", "devfs", "iso9660", "overlay", "aufs", "squashfs"]
```

/etc/telegraf/telegraf.d/ping-telegraf.conf
```
[[inputs.ping]]
    count = 30
    ping_interval = 2.0
    timeout = 0.9
    deadline = 58
    urls = [
    "localhost",
    "gateway"
]
```

ClickHouse:
```
SHOW CREATE TABLE telegraf3.ping
 CREATE TABLE telegraf3.ping
(
    `timestamp` DateTime,
    `host` String,
    `url` String,
    `standard_deviation_ms` Float64,
    `packets_transmitted` Int32,
    `packets_received` Int32,
    `percent_packet_loss` Float64,
    `ttl` Int32,
    `minimum_response_ms` Float64,
    `maximum_response_ms` Float64,
    `result_code` Int32,
    `average_response_ms` Float64
)
ENGINE = MergeTree
PARTITION BY toYYYYMMDD(timestamp)
PRIMARY KEY (timestamp, host)
ORDER BY (timestamp, host)
SETTINGS index_granularity = 8192

--

CREATE TABLE telegraf3.disk
(
    `timestamp` DateTime,
    `device` String,
    `fstype` String,
    `host` String,
    `mode` String,
    `path` String,
    `used_percent` Float64,
    `inodes_total` UInt32,
    `inodes_free` UInt32,
    `inodes_used` UInt32,
    `total` UInt32,
    `free` UInt32,
    `used` UInt32
)
ENGINE = MergeTree
PARTITION BY toYYYYMMDD(timestamp)
PRIMARY KEY (timestamp, host)
ORDER BY (timestamp, host)
SETTINGS index_granularity = 8192

--

CREATE TABLE telegraf3.cpu
(
    `timestamp` DateTime,
    `cpu` String,
    `host` String,
    `usage_system` Float64,
    `usage_iowait` Float64,
    `usage_irq` Float64,
    `usage_steal` Float64,
    `usage_guest_nice` Float64,
    `usage_user` Float64,
    `usage_idle` Float64,
    `usage_nice` Float64,
    `usage_softirq` Float64,
    `usage_guest` Float64
)
ENGINE = MergeTree
PARTITION BY toYYYYMMDD(timestamp)
PRIMARY KEY (timestamp, host)
ORDER BY (timestamp, host)
SETTINGS index_granularity = 8192
```

```
SELECT *
FROM telegraf3.ping
WHERE (url = 'localhost') 
LIMIT 1

┌───────────timestamp─┬─host─┬─url───────┬─standard_deviation_ms─┬─packets_transmitted─┬─packets_received─┬─percent_packet_loss─┬─ttl─┬─minimum_response_ms─┬─maximum_response_ms─┬─result_code─┬─average_response_ms─┐
│ 2022-12-11 00:00:50 │ host │ localhost │                 0.155 │                  30 │               30 │                   0 │  64 │               0.021 │               1.435 │           1 │               0.871 │
└─────────────────────┴──────┴───────────┴───────────────────────┴─────────────────────┴──────────────────┴─────────────────────┴─────┴─────────────────────┴─────────────────────┴─────────────┴─────────────────────┘

```

изображение

**Comment by Okaho** [comment]
@zentkhv, did you succeed?

**Comment by zentkhv** [comment]
@Okaho , How did you understand which fields need to be added to the clickhouse table structure?

**Comment by Okaho** [comment]
My tables were created by telegraf, the above structures are given as an example. I suggest deleting the tables. The telegraf version is very important, i have 1.24.0

**Comment by zentkhv** [comment]
@Okaho , thank you very much! Finally I managed to write the data. I have a telegraph version 1.25.0, it writes both to a pre-created table, and can create its own.

**Comment by Okaho** [comment]
Great, very glad it worked out for you!

**Comment by Okaho** [comment]
Help me set up a route to send data to different systems. Using "routing plugin" https://www.influxdata.com/blog/telegraf-best-practices/ data stops going to both output plugin.

**Comment by majlsfiles** [comment]
> @Okaho , thank you very much! Finally I managed to write the data. I have a telegraph version 1.25.0, it writes both to a pre-created table, and can create its own.

I am using telegraf 1.29.4 and ClickHouse 24.1.5.6, while everything works great I have problem that metrics are written slow and after a while buffer is full and metrics are dropped.

2024-02-21T09:31:04+01:00 D! [outputs.sql] Wrote batch of 1000 metrics in 6.234639174s

This should probably be in ms and not in seconds. Can someone please help with this issue?

**Pull request merged**